src/Xamarin.Android.Build.Tasks/Tasks/LinkApplicationSharedLibraries.cs

```diff
@@ -0,0 +1,266 @@
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Threading;
+using ThreadingTasks = System.Threading.Tasks;
+
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+
+using Xamarin.Android.Tools;
+using Xamarin.Build;
+
+namespace Xamarin.Android.Tasks
+{
+	public class LinkApplicationSharedLibraries : AsyncTask
+	{
+		sealed class Config
+		{
+			public string LinkerPath;
+			public string LinkerOptions;
+			public string OutputSharedLibrary;
+		}
+
+		sealed class InputFiles
+		{
+			public List<string> ObjectFiles;
+			public string OutputSharedLibrary;
+		}
+
+		[Required]
+		public ITaskItem[] ObjectFiles { get; set; }
+
+		[Required]
+		public ITaskItem[] ApplicationSharedLibraries { get; set; }
+
+		[Required]
+		public bool DebugBuild { get; set; }
+
+		[Required]
+		public string AndroidNdkDirectory { get; set; }
+
+		public override bool Execute ()
+		{
+			try {
+				return DoExecute ();
+			} catch (Exception e) {
+				// We don't want to present a wall of text to the end user...
+				Log.LogCodedError ("XA3001", $"Unable to link `libxamarin-app.so`: {e.Message}");
+
+				// ...but having the full stack trace available is important for us, so log it here
+				// int the way that will make it included in XA builds on bots or in verbose end user
+				// builds but not otherwise.
+				LogDebugMessage ($"Full exception: {e}");
+				return false;
+			}
+		}
+
+		bool DoExecute ()
+		{
+			Yield ();
+			try {
+				var task = ThreadingTasks.Task.Run ( () => RunParallelLinker (), CancellationToken);
+				task.ContinueWith (Complete);
```

Comment:
Example of using new extension method: https://github.com/xamarin/xamarin-android/blob/0d211ceb5c21f9f47d7894c1400596052055fb0c/src/Xamarin.Android.Build.Tasks/Tasks/Aapt.cs#L219-L221